### DISCUSSION TITLE
Timeout of queued messages

### DISCUSSION THREAD

**Discussion by peterdstone** [relayrides/pushy#1022] [Q&A]
We're using Pushy in both a async and synchronous mode and want to be able to handle connection issues for the synchronous messages as w recently ran out of threads.

Part of our solution is going to be to add a timeout to the get to prevent our threads locking but from other questions it seems cancelling the future doesn't cancel the message to be sent.

Could I confirm this statement, and if we can't cancel message using the future do we have any control on how long a message should be queued before timing out.

**Comment by jchambers** [comment]
> it seems cancelling the future doesn't cancel the message to be sent.

This is correct; I suppose we could make futures cancellable up to the point where they're passed to the OS network stack, but that's often a very short window in practice.

> …do we have any control on how long a message should be queued before timing out.

In the sense I think you're asking about, no. But please let me refer you to the [best practices](https://github.com/jchambers/pushy/wiki/Best-practices) page on the wiki, and the ["flow control" section](https://github.com/jchambers/pushy/wiki/Best-practices#flow-control) in particular. It may be that your use case is best served by using something between your existing application layer and Pushy to control the rate at which you're sending (and possibly enqueuing) notifications.

**Reply by peterdstone** [reply]
Thanks for confirming.
I think our issues stemed from exceptional events, it may have been cert issues, causing the synchronous threads to stick around for days. We may have other issues but certainly we need to release our thread by putting a timeout on the futures get.
Our throughput isn't high so we don't currently have our own flow control, but it's somthing to look at.
